### PR TITLE
feat: add collections to data store

### PR DESCRIPTION
**Pull request by AjimenezDCL**
Our only data-driven implementation is based on `ScriptableObjects`. It has been marked as `Legacy` and a more consistent one has been implemented.

The new set of `BaseCollection` and `BaseDictionary` should allow you to use them in our `DataStore`. They also contain more methods/events and now you can access the whole collections through an `IEnumerable` for integrity reasons.

### CODE REVIEW

File: unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return false;
+
+        OnRemoved?.Invoke(element);
+        return true;
+    }
+
+    public void RemoveAt(int index)
+    {
+        list.RemoveAt(index);
+    }
```
Comment (pbosio) [inline unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs]: question: shouldn't `RemoveAt` also raise `OnRemoved` event?

File: unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Collections.Generic;
+
+public class BaseCollection<T> : IBaseCollection<T>, IEquatable<IEnumerable<T>>
+{
+    public event Action<IEnumerable<T>> OnChange;
```
Comment (pbosio) [inline unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs]: I feel `OnChange` naming a little misleading here cause removing and/or adding doesn't trigger this event even if in some way the collection has changed. I can't come up with any suggestion right now tho

File: unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Collections.Generic;
+
+public class BaseDictionary<TKey, TValue> : IBaseDictionary<TKey, TValue>
+{
+    public event Action<IEnumerable<KeyValuePair<TKey, TValue>>> OnChange;
```
Comment (pbosio) [inline unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseDictionary.cs]: same comment as with `BaseCollection.cs` `OnChange` event name

**Review by pbosio — DISMISSED** [review]

File: unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs (hunk 1 above)
Comment (AjimenezDCL) [reply unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs]: Dammit, at the beggining was just a wrapper to other function that dissapeared. Gonna add it, nice catch!

File: unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs (hunk 2 above)
Comment (AjimenezDCL) [reply unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs]: What do you think of `OnSet`?

Comment (pbosio) [reply unity-client/Assets/Scripts/MainScripts/DCL/DataStore/Variables/BaseCollection.cs]: 👍  I like that name

**Review by sandrade-dcl — APPROVED** [review]
Good work! :)

**Review by pravusjif — APPROVED** [review]

**Pull request merged**